gprofiler/spark/spark_collector.py

```diff
@@ -0,0 +1,992 @@
+import contextlib
+import json
+import os
+import re
+import threading
+import time
+import traceback
+import xml.etree.ElementTree as ET
+from abc import ABC, abstractmethod
+from datetime import datetime, timedelta
+from typing import Any, Callable, Dict, Generator, List, Optional, Tuple, Union, cast
+from urllib.parse import urljoin, urlparse
+
+import psutil
+import requests
+from granulate_utils.linux.process import is_process_running
+from psutil import AccessDenied, NoSuchProcess, Process, process_iter
+from requests.exceptions import ConnectionError, HTTPError, InvalidURL, Timeout
+
+import gprofiler.spark.metrics as metrics_definition
+from gprofiler.log import get_logger_adapter
+from gprofiler.metadata.system_metadata import get_hostname
+from gprofiler.platform import is_windows
+from gprofiler.spark.client import SparkAPIClient
+from gprofiler.utils import get_iso8601_format_time
+
+GMT_TIME_FORMAT = "%Y-%m-%dT%H:%M:%S.%fGMT"
+
+# Application type and states to collect
+YARN_SPARK_APPLICATION_SPECIFIER = "SPARK"
+YARN_RUNNING_APPLICATION_SPECIFIER = "RUNNING"
+YARN_COMPLETED_APPLICATION_SPECIFIER = "FINISHED"
+
+# SPARK modes
+SPARK_DRIVER_MODE = "driver"
+SPARK_YARN_MODE = "yarn"
+SPARK_MESOS_MODE = "mesos"
+
+# COMMON urls
+YARN_APPS_PATH = "ws/v1/cluster/apps"
+YARN_CLUSTER_PATH = "ws/v1/cluster/metrics"
+YARN_NODES_PATH = "ws/v1/cluster/nodes"
+SPARK_APPS_PATH = "api/v1/applications"
+MESOS_MASTER_APP_PATH = "/frameworks"
+
+# METRIC keys
+METRIC_TIMESTAMP_KEY = "timestamp"
+METRICS_DATA_KEY = "metrics"
+METRICS_FILE_PREFIX = "spark_metric_"
+
+STRUCTURED_STREAMS_METRICS_REGEX = re.compile(
+    r"^[\w-]+\.driver\.spark\.streaming\.(?P<query_name>[\w-]+)\.(?P<metric_name>[\w-]+)$"
+)
+
+
+class BaseCollector(ABC):
```

Comment:
Since it's only used in `SparkCollector` we can remove this generalization for now (i.e `SparkCollector` needn't inherit from this class, and we can remove it)

Reply:
I removed BaseCollector class, and moved code of collect() to SparkCollector